src/shared/mount-util.c

```diff
@@ -742,3 +745,266 @@ int mount_option_mangle(
+
+        assert(target > 0);
+        assert(propagate_path);
+        assert(incoming_path);
+        assert(src);
+        assert(dest);
+
+        /* One day, when bind mounting /proc/self/fd/n works across
+         * namespace boundaries we should rework this logic to make
+         * use of it... */
+
+        p = strjoina(propagate_path, "/");
+        if (laccess(p, F_OK) < 0) {
+                if (error_message)
+                        *error_message = strdup("Target does not allow propagation of mount points");
+                return -EOPNOTSUPP;
+        }
+
+        r = chase_symlinks(src, NULL, CHASE_TRAIL_SLASH, &chased_src, NULL);
+        if (r < 0) {
+                if (error_message)
+                        *error_message = strdup("Failed to resolve source path");
+                return r;
+        }
+
+        if (lstat(chased_src, &st) < 0) {
+                r = -errno;
+                if (error_message)
+                        *error_message = strdup("Failed to stat() source path");
+                return r;
+        }
+        if (S_ISLNK(st.st_mode)) /* This shouldn't really happen, given that we just chased the symlinks above, but let's better be safe… */ {
+                if (error_message)
+                        *error_message = strdup("Source directory can't be a symbolic link");
+                return -EOPNOTSUPP;
+        }
+
+        /* Our goal is to install a new bind mount into the container,
+           possibly read-only. This is irritatingly complex
+           unfortunately, currently.
+
+           First, we start by creating a private playground in /tmp,
+           that we can mount MS_SLAVE. (Which is necessary, since
+           MS_MOVE cannot be applied to mounts with MS_SHARED parent
+           mounts.) */
+
+        if (!mkdtemp(mount_slave)) {
+                r = -errno;
+                if (error_message)
+                        *error_message = strjoin("Failed to create playground ", mount_slave);
+                return r;
+        }
+
+        mount_slave_created = true;
+
+        r = mount_nofollow_verbose(LOG_DEBUG, mount_slave, mount_slave, NULL, MS_BIND, NULL);
+        if (r < 0) {
+                if (error_message)
+                        *error_message = strjoin("Failed to make bind mount ", mount_slave);
+                goto finish;
+        }
+
+        mount_slave_mounted = true;
+
+        r = mount_nofollow_verbose(LOG_DEBUG, NULL, mount_slave, NULL, MS_SLAVE, NULL);
+        if (r < 0) {
+                if (error_message)
+                        *error_message = strjoin("Failed to remount slave ", mount_slave);
+                goto finish;
+        }
+
+        /* Second, we mount the source file or directory to a directory inside of our MS_SLAVE playground. */
+        mount_tmp = strjoina(mount_slave, "/mount");
+        if (S_ISDIR(st.st_mode))
+                r = mkdir_errno_wrapper(mount_tmp, 0700);
+        else
+                r = touch(mount_tmp);
```

Comment:
one of these days we should add a helper call to mount-util.h: 

```c
int make_mount_point_inode(const char *path, mode_t mode);
```

i.e. that does what these 4 lines do in a nicely abstracted way. I am pretty sure we have the same 4 lines at other places too

Reply:
May I do that as a separate followup? Will fix other instances as well

Reply:
https://github.com/systemd/systemd/pull/18030